File: source/bindings/csharp/pronunciation_assessment_config.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using System;
using System.Globalization;
using Microsoft.CognitiveServices.Speech.Internal;
using static Microsoft.CognitiveServices.Speech.Internal.SpxExceptionThrower;

namespace Microsoft.CognitiveServices.Speech.PronunciationAssessment
{
    /// <summary>
    /// Class that defines pronunciation assessment configuration
    /// Added in 1.14.0
    /// </summary>
    public class PronunciationAssessmentConfig
    {
        internal InteropSafeHandle configHandle;
        internal PropertyCollection progBag = null;

        internal PronunciationAssessmentConfig(IntPtr configPtr)
        {
            Setup(configPtr);
        }

        /// <summary>
        /// Creates an instance of the PronunciationAssessmentConfig
        /// For the parameters details, see
        /// https://docs.microsoft.com/azure/cognitive-services/speech-service/rest-speech-to-text#pronunciation-assessment-parameters
        /// </summary>
        /// <param name="referenceText">The reference text</param>
        /// <param name="gradingSystem">The point system for score calibration</param>
        /// <param name="granularity">The evaluation granularity</param>
        /// <param name="enableMiscue">If enables miscue calculation</param>
        /// <returns>A new PronunciationAssessmentConfig instance.</returns>
        public PronunciationAssessmentConfig(string referenceText,
            GradingSystem gradingSystem = GradingSystem.FivePoint,
            Granularity granularity = Granularity.Phoneme,
            bool enableMiscue = false)
        {
            IntPtr pronunciationAssessmentConfigHandle = IntPtr.Zero;
            IntPtr referenceTexTPtr = Utf8StringMarshaler.MarshalManagedToNative(referenceText);
            ThrowIfFail(Internal.PronunciationAssessmentConfig.create_pronunciation_assessment_config(out pronunciationAssessmentConfigHandle, referenceTexTPtr, (int)gradingSystem, (int)granularity, enableMiscue));
            Setup(pronunciationAssessmentConfigHandle);
        }

        internal void Setup(IntPtr configPtr)
        {
            ThrowIfNull(configPtr);
            configHandle = new InteropSafeHandle(configPtr, Internal.PronunciationAssessmentConfig.pronunciation_assessment_config_release);
            IntPtr hpropbag = IntPtr.Zero;
            ThrowIfFail(Internal.PronunciationAssessmentConfig.pronunciation_assessment_config_get_property_bag(configHandle, out hpropbag));
            progBag = new PropertyCollection(hpropbag);
        }

        /// <summary>
        /// Creates an instance of the PronunciationAssessmentConfig from json.
        /// see https://docs.microsoft.com/azure/cognitive-services/speech-service/rest-speech-to-text#pronunciation-assessment-parameters
        /// </summary>
        /// <param name="json">The json string containing the pronunciation assessment parameters.</param>
        /// <returns>A new PronunciationAssessmentConfig instance.</returns>
        public static PronunciationAssessmentConfig FromJson(string json)
        {
            IntPtr pronunciationAssessmentConfigHandle = IntPtr.Zero;
            IntPtr jsonPtr = Utf8StringMarshaler.MarshalManagedToNative(json);
            ThrowIfFail(Internal.PronunciationAssessmentConfig.create_pronunciation_assessment_config_from_json(out pronunciationAssessmentConfigHandle, jsonPtr));
            return new PronunciationAssessmentConfig(pronunciationAssessmentConfigHandle);
        }

        /// <summary>
        /// Reference Text
        /// </summary>
        public string ReferenceText
        {
            get
            {
                return progBag.GetProperty(PropertyId.PronunciationAssessment_ReferenceText);
            }

            set
            {
                progBag.SetProperty(PropertyId.PronunciationAssessment_ReferenceText, value);
            }
        }

        /// <summary>
        /// Gets and Sets the scenario ID, which is a GUID indicating a customized point system.
        /// </summary>
        public string ScenarioId
        {
            get
            {
                return progBag.GetProperty(PropertyId.PronunciationAssessment_ScenarioId);
            }

            set
            {
                progBag.SetProperty(PropertyId.PronunciationAssessment_ScenarioId, value);
            }
        }

        /// <summary>
        /// Gets to json string of pronunciation assessment parameters.
        /// </summary>
        /// <returns>json string of pronunciation assessment parameters.</returns>
        public string ToJson()
        {
            ThrowIfNull(configHandle);
            string jsonVal = string.Empty;
            IntPtr pStr = Internal.PronunciationAssessmentConfig.pronunciation_assessment_config_to_json(configHandle);
            if (pStr != IntPtr.Zero)
            {
                try
                {
                    jsonVal = Utf8StringMarshaler.MarshalNativeToManaged(pStr);
                }
                finally
                {
                    ThrowIfFail(Internal.PropertyCollection.property_bag_free_string(pStr));
                }
            }

            return jsonVal;
        }

        /// <summary>
        /// Applies the settings in this config to a recognizer.
        /// </summary>
        /// <param name="recognizer">The target recognizer.</param>
        public void ApplyTo(Recognizer recognizer)
        {
            ThrowIfNull(configHandle);
            ThrowIfNull(recognizer);
            ThrowIfNull(recognizer.recoHandle);
            ThrowIfNull(Internal.PronunciationAssessmentConfig.pronunciation_assessment_config_apply_to_recognizer(configHandle, recognizer.recoHandle));
        }
    }
}

File: source/bindings/java/com/microsoft/cognitiveservices/speech/PronunciationAssessmentConfig.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
package com.microsoft.cognitiveservices.speech;

import java.io.Closeable;
import java.util.ArrayList;
import java.util.List;
import com.microsoft.cognitiveservices.speech.SpeechConfig;
import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.util.IntRef;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;
import com.microsoft.cognitiveservices.speech.util.SafeHandleType;
import com.microsoft.cognitiveservices.speech.PronunciationAssessmentGradingSystem;
import com.microsoft.cognitiveservices.speech.PronunciationAssessmentGranularity;
import com.microsoft.cognitiveservices.speech.Recognizer;

/**
 * Represents pronunciation assessment configuration
 * Added in version 1.14.0
 */
public final class PronunciationAssessmentConfig implements Closeable
{
    // load the native library.
    static {
        // trigger loading of native library
        try {
            Class.forName(SpeechConfig.class.getName());
        }
        catch (ClassNotFoundException ex) {
            throw new IllegalStateException(ex);
        }
    }

    /**
     * Initializes an instance of the PronunciationAssessmentConfig
     * For the parameters details, see
     * https://docs.microsoft.com/azure/cognitive-services/speech-service/rest-speech-to-text#pronunciation-assessment-parameters
     * @param referenceText The reference text
     * @param gradingSystem The point system for score calibration
     * @param granularity The evaluation granularity
     * @param enableMiscue If enables miscue calculation
     */
    public PronunciationAssessmentConfig(String referenceText, PronunciationAssessmentGradingSystem gradingSystem, PronunciationAssessmentGranularity granularity, boolean enableMiscue) {
        IntRef pronAssessmentConfigRef = new IntRef(0);
        Contracts.throwIfFail(create(pronAssessmentConfigRef, referenceText, gradingSystem.getValue(), granularity.getValue(), enableMiscue));
        init(pronAssessmentConfigRef.getValue());
    }

    /**
     * Initializes an instance of the PronunciationAssessmentConfig
     * For the parameters details, see
     * https://docs.microsoft.com/azure/cognitive-services/speech-service/rest-speech-to-text#pronunciation-assessment-parameters
     * @param referenceText The reference text
     * @param gradingSystem The point system for score calibration
     * @param granularity The evaluation granularity
     */
    public PronunciationAssessmentConfig(String referenceText, PronunciationAssessmentGradingSystem gradingSystem, PronunciationAssessmentGranularity granularity) {
        this(referenceText, gradingSystem, granularity, false);
    }

    /**
     * Initializes an instance of the PronunciationAssessmentConfig
     * For the parameters details, see
     * https://docs.microsoft.com/azure/cognitive-services/speech-service/rest-speech-to-text#pronunciation-assessment-parameters
     * @param referenceText The reference text
     */
    public PronunciationAssessmentConfig(String referenceText) {
        this(referenceText, PronunciationAssessmentGradingSystem.FivePoint, PronunciationAssessmentGranularity.Phoneme);
    }

    /**
     * Creates an instance of the PronunciationAssessmentConfig from a json string
     * @param json the json string
     * @return The pronunciation assessment configuration being created.
     */
    public static PronunciationAssessmentConfig fromJson(String json) {
        IntRef pronAssessmentConfigRef = new IntRef(0);
        Contracts.throwIfFail(fromJson(pronAssessmentConfigRef, json));
        return new PronunciationAssessmentConfig(pronAssessmentConfigRef.getValue());
    }

    /**
     * Explicitly frees any external resource attached to the object
     */
    @Override
    public void close() {
        if (disposed) {
            return;
        }
        if (propertyHandle != null)
        {
            propertyHandle.close();
            propertyHandle = null;
        }
        if (configHandle != null){
            this.configHandle.close();
            this.configHandle = null;
        }
        disposed = true;
    }

    /**
     * Sets the reference text.
     * @param value the reference text.
     */
    public void setReferenceText(String value) {
        Contracts.throwIfNull(value, "value");
        propertyHandle.setProperty(PropertyId.PronunciationAssessment_ReferenceText, value);
    }

    /**
     * Gets the reference text.
     * @return The reference text.
     */
    public String getReferenceText() {
        return propertyHandle.getProperty(PropertyId.PronunciationAssessment_ReferenceText);
    }

    /**
     * Gets the json string of pronunciation assessment parameters.
     * @return json string of pronunciation assessment parameters.
     */
    public String toJson() {
        Contracts.throwIfNull(configHandle, "configHandle");
        return toJson(configHandle);
    }

    /**
     * Applies the settings in this config to a recognizer.
     * @param recognizer The target recognizer.
     */
    public void applyTo(Recognizer recognizer)
    {
        Contracts.throwIfNull(configHandle, "configHandle");
        Contracts.throwIfNull(recognizer, "recognizer");
        Contracts.throwIfNull(recognizer.getImpl(), "recoHandle");
        applyTo(configHandle, recognizer.getImpl());
    }

    /*! \cond INTERNAL */
    /**
     * Returns the PronunciationAssessmentConfig
     * @return The implementation of the PronunciationAssessmentConfig object
     */
    public SafeHandle getImpl()
    {
        return configHandle;
    }
    /*! \endcond */

    private PronunciationAssessmentConfig(long handleValue) {
        init(handleValue);
    }

    private void init(long handleValue) {
        Contracts.throwIfNull(handleValue, "handleValue");
        this.configHandle = new SafeHandle(handleValue, SafeHandleType.PronunciationAssessmentConfig);
        IntRef propHandle = new IntRef(0);
        Contracts.throwIfFail(getPropertyBag(configHandle, propHandle));
        this.propertyHandle = new PropertyCollection(propHandle);
    }

    private final static native long create(IntRef pronConfigRef, String referenceText, int gradingSystem, int granularity, boolean enableMiscue);
    private final static native long fromJson(IntRef pronConfigRef, String json);
    private final native long getPropertyBag(SafeHandle pronConfigRef, IntRef propHandle);
    private final native long applyTo(SafeHandle pronConfigRef, SafeHandle recoHandle);
    private final native String toJson(SafeHandle pronConfigRef);

    private SafeHandle configHandle = null;
    private PropertyCollection propertyHandle = null;
    private boolean disposed = false;
}
